test/framework/src/main/java/org/elasticsearch/transport/MockTcpTransport.java

```diff
@@ -318,6 +303,7 @@ public void accept(Executor executor) throws IOException {
                 MockChannel incomingChannel = null;
                 try {
                     configureSocket(incomingSocket);
+                    incomingSocket.setSoLinger(true, 0);
```

Comment:
I am hesitating to set a 0 timeout for `SO_LINGER` The reason is that the abnormal termination of a socket that causes it to move into `TIME_WAIT` can be caused due to an exception ie. during request parsing on the server then the server closes the connection and in this case we should really stay in `TIME_WAIT`. The other case where the server closes the connection instead of the client is when we shutdown and then we can set our socket into `SO_LINGER=0` since it's the right thing todo in this situation. I actually think we should do this in a broader scope. There is for instance a method `TcpTransport#closeChannels(List<Channels> channels, boolean blocking)` I wonder if we can piggyback on this boolean and set `SO_LINGER` to `true,0` if we close our transport. It might even be a better idea to add another boolean `boolean closingTransport` in this case we can do the right thing also in other transport impls?

Reply:
@s1monw 

> It might even be a better idea to add another boolean boolean closingTransport in this case we can do the right thing also in other transport impls?

Makes sense to me to do the additional `boolean`. My reasoning would be that the `blocking` parameter currently seems to be more about the way handle Netty's internal threading and not any low- level `Socket` behavior/setting.

Reply:
@s1monw done, added another parameter and only made it `RST` on `org.elasticsearch.transport.TcpTransport#doStop` and the actual `close` call of `TcpTransport` leaving the failure handlers and such untouched.
This still (after removing the SO_LINGER setting I added) resolves the issue just fine, tests still don't leak any `TIME_WAIT`s.
Hope this is what you had in mind :)